### PR TITLE
[spirv] Support SPV_EXT_demote_to_helper_invocation

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

**Comment by ehsannas** [comment]
The SPIR-V backend used to generate `OpKill` for the HLSL `discard` statement ([reference](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-discard)). However, this is not quite correct. The proper behavior is to use `OpDemoteToHelperInvocationEXT` ([reference](https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/EXT/SPV_EXT_demote_to_helper_invocation.asciidoc))

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp [hunk 1]
```diff
@@ -1398,12 +1398,18 @@ spv::LoopControlMask SpirvEmitter::translateLoopAttribute(const Stmt *stmt,
...
   assert(!spvBuilder.isCurrentBasicBlockTerminated());
-  spvBuilder.createKill(discardStmt->getLoc());
-  // Some statements that alter the control flow (break, continue, return, and
-  // discard), require creation of a new basic block to hold any statement that
-  // may follow them.
-  auto *newBB = spvBuilder.createBasicBlock();
-  spvBuilder.setInsertPoint(newBB);
+
+  // The discard statement can only be called from a pixel shader
+  assert(spvContext.isPS());
```
Comment (jaebaek) [inline tools/clang/lib/SPIRV/SpirvEmitter.cpp]: I am not sure what is the expected behavior from the AST parser, but I think we should print an explicit error message saying "The discard statement can only be called from a pixel shader".

I am just worried what if a user use release build and is not able to recognize it.

File: tools/clang/test/CodeGenSPIRV/cf.discard.hlsl [hunk 2]
```diff
@@ -1,7 +1,9 @@
 // Run: %dxc -T ps_6_0 -E main
```
Comment (jaebaek) [inline tools/clang/test/CodeGenSPIRV/cf.discard.hlsl]: Why don't we add a test that uses `discard;` in a vertex shader and check it with the expected error?

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp (hunk 1 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp]: Good idea. Done.

File: tools/clang/test/CodeGenSPIRV/cf.discard.hlsl (hunk 2 above)
Comment (ehsannas) [reply tools/clang/test/CodeGenSPIRV/cf.discard.hlsl]: Done

**Pull request merged**